### PR TITLE
Sync tree: Delete translations and alias pages

### PR DESCRIPTION
**Pull request by KalobTaulien**
This is part of #7592. Builds on #7984

When an origin page is awaiting confirmation to be deleted, check for translated pages and alias pages to provide a better delete message in the confirm_delete.html template. 

When an origin page is deleted, get all the translated pages (and get the alias pages at the same time), check to see if each of those pages share the same parent page and if so then delete the page using `DeletePageAction()` rather than a pure `.delete()`. 

Before merging into `main` we still need the move pages, have bulk actions, and supply tests.

### CODE REVIEW

File: wagtail/admin/views/pages/delete.py [hunk 1]
```diff
@@ -17,32 +17,59 @@ def delete(request, page_id):
...
         next_url = get_valid_next_url_from_request(request)
+        parent_page_translations = page.get_parent().get_translations()
+        translations_to_delete = [
+            page
+            for page in page.get_translations()
+            if page.get_parent() in parent_page_translations
+        ]
+        translation_descendant_count = sum(
+            [p.get_descendants().count() for p in translations_to_delete]
+        )
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:34]: @KalobTaulien https://github.com/wagtail/wagtail/issues/7592#issuecomment-1034790403 states we will use a hook to gather this data and then the various actions will act upon it. 
The hook only runs when `WAGTAIL_I18N_ENABLED = True`, and for core we would implent this in simple_translation only when `WAGTAILSIMPLETRANSLATION_SYNC_PAGE_TREE = True`.

We discussed that https://github.com/wagtail/wagtail/pull/7984/files lays the foundation and you'be building on top of it.

**Review by zerolab — CHANGES_REQUESTED** [review]

File: wagtail/admin/views/pages/delete.py [hunk 2]
```diff
@@ -17,32 +18,66 @@ def delete(request, page_id):
...
+        if request.method == "POST":
             parent_id = page.get_parent().id
+            # Delete the source page.
             action = DeletePageAction(page, user=request.user)
             # Permission checks are done above, so skip them in execute.
             action.execute(skip_permission_checks=True)
 
-            messages.success(request, _("Page '{0}' deleted.").format(page.get_admin_display_title()))
+            # Delete translation and alias pages if they have the same parent page.
+            parent_page_translations = page.get_parent().get_translations()
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:49]: this should only kick in when `WAGTAIL_I18N_ENABLED = True`

File: wagtail/admin/views/pages/delete.py [hunk 3]
```diff
@@ -17,32 +18,66 @@ def delete(request, page_id):
...
+                    action = DeletePageAction(_page, user=request.user)
+                    # Permission checks are done above, so skip them in execute.
+                    action.execute(skip_permission_checks=True)
+
+            messages.success(
+                request, _("Page '{0}' deleted.").format(page.get_admin_display_title())
+            )
 
-            for fn in hooks.get_hooks('after_delete_page'):
+            for fn in hooks.get_hooks("after_delete_page"):
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:60]: thinking out loud: what happens if the site has a custom `after_delete_page` hook implementation that does something specific (e.g. notifies an admin). 

We call this for the source page as before, which is great. Should we call the hook for the translated pages? cc @phildexter

File: wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html [hunk 4]
```diff
@@ -9,11 +9,29 @@
     <div class="nice-padding">
         <p>
             {% trans 'Are you sure you want to delete this page?' %}
-            {% if descendant_count %}
+
+            {% if descendant_count and translation_count %}
+                {% blocktrans count counter=descendant_count %}
+                    Deleting this page will also delete {{ descendant_count }} subpage,
```
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html:15]: sorry @KalobTaulien, just updated the messages for delete in https://github.com/wagtail/wagtail/issues/7592#issuecomment-1034790403 -- we're trying to standardise on "child pages" rather than subpages. Not sure how I missed updating this ealier (Unpublish/Move) have the correct messages 🤦‍♂️

File: wagtail/admin/views/pages/delete.py (hunk 3 above)
Comment (KalobTaulien) [reply wagtail/admin/views/pages/delete.py:60]: That's a great question. I feel like, yes, we should call the hook for translated pages. Also thinking out loud here: 

Would we want to create a _new_ i18n enabled hook or just pass all the translated pages into the existing hook? If we pass it into the existing `after_delete_page` hook, would that break past code? If we created its own i18n enabled hook, we are creating further divergence in hooks, no? Though if we had a bunch of i18n-only hooks it could make sense to create them now (or soon..ish).

File: wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html (hunk 4 above)
Comment (KalobTaulien) [reply wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html]: No problem! I updated these, let me know if I missed anything in the new copy though.

File: wagtail/admin/views/pages/delete.py [hunk 5]
```diff
@@ -17,32 +18,67 @@ def delete(request, page_id):
...
             parent_id = page.get_parent().id
+            # Delete the source page.
             action = DeletePageAction(page, user=request.user)
             # Permission checks are done above, so skip them in execute.
             action.execute(skip_permission_checks=True)
 
-            messages.success(request, _("Page '{0}' deleted.").format(page.get_admin_display_title()))
+            # Delete translation and alias pages if they have the same parent page.
+            if getattr(settings, 'WAGTAIL_I18N_ENABLED', False):
+                parent_page_translations = page.get_parent().get_translations()
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:50]: I still feel this should move to the simple_translation hook implemnentation.

tl;dr the delete view becomes a simple actor on the list of pages that need action. Should a developer decide to provide more, then that's ok

File: wagtail/admin/views/pages/delete.py (hunk 3 above)
Comment (phildexter) [reply wagtail/admin/views/pages/delete.py:61]: Not sure really. I think if I'm having an opinion, if it's an alias page then don't call it (because the page didn't really exist as a normal page) but if it's a translated page then do call it...does that feel consistent with the other things we're doing?

Comment (zerolab) [reply wagtail/admin/views/pages/delete.py:61]: Let's call it for all, but only handle redirects for the page the delete was called on.

Comment (KalobTaulien) [reply wagtail/admin/views/pages/delete.py]: @zerolab @phildexter Redirects are only automatically created for `move` actions and when a slug is changed. I can add that in #8032 though. If we agree to do that I'll add a todo on that PR so we don't forget to add it.

Comment (zerolab) [reply wagtail/admin/views/pages/delete.py]: redirects in this case was probablu a bit misleading. it refers to 

```python
            for fn in hooks.get_hooks("after_delete_page"):
                result = fn(request, page)
                if hasattr(result, "status_code"):
                    return result
```

i.e. if the hook returns a result that has a status_code (usually a `HttpResponse`), then this returns. 

what I meant is call the after hook for the other pages, but don't return anything

Comment (KalobTaulien) [reply wagtail/admin/views/pages/delete.py:61]: Updated here: https://github.com/wagtail/wagtail/pull/8002/commits/5ba40c00b2fcc56345b0d41453d30d561a29c6ea 

Let me know if this _isn't_ what we were thinking.

Comment (easherma-truth) [reply wagtail/admin/views/pages/delete.py:61]: https://github.com/wagtail/wagtail-personalisation/blob/48955675be73ba0753229660aae9c774c6f65bed/src/wagtail_personalisation/wagtail_hooks.py#L304

Might be another relevant case to consider. As well as the fact that wagtail_personalization will prevent pages from being deleted if they are canonical: 

https://github.com/wagtail/wagtail-personalisation/blob/137b5b411ccbe67cca2cc84fe1e7f63c78e8133e/src/wagtail_personalisation/models.py#L215

So I'm curious how those sorts of things would interact with this? 

That being said, I think it's overall a great idea! (They said as they spent plenty of time yesterday manually deleting pages and their translations in the shell the other day...)

File: wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html [hunk 6]
```diff
@@ -9,11 +9,29 @@
     <div class="nice-padding">
         <p>
             {% trans 'Are you sure you want to delete this page?' %}
-            {% if descendant_count %}
+
+            {% if descendant_count and translation_count %}
+                {% blocktrans count counter=descendant_count %}
```
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html:14]: please use `blocktrans trimmed` everywhere

ref: #7113

File: wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html [hunk 7]
```diff
@@ -9,11 +9,29 @@
...
+                    Deleting this page will also delete {{ descendant_count }} of their child pages,
+                {% plural %}
+                    Deleting this page will also delete {{ descendant_count }} more child pages,
```
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html:17]: breaking the sentence on a comma like this breaks the translator context.

What about:

```jinja
{% blocktrans trimmed count counter=translation_count %}
Deleting this page will also delete {{ descendant_count }} more child pages, {{ translation_count }} translation of this page and their combined {{ combined_subpages }} translated child pages.
{% plural %}
Deleting this page will also delete {{ descendant_count }} more child pages, {{ translation_count }} translations of this page and their combined {{ combined_subpages }} translated child pages.
{% endblocktrans %}
```

File: wagtail/admin/views/pages/delete.py [hunk 8]
```diff
@@ -24,17 +25,44 @@ def delete(request, page_id):
...
+                        action.execute(skip_permission_checks=True)
+
             messages.success(
                 request, _("Page '{0}' deleted.").format(page.get_admin_display_title())
             )
 
             for fn in hooks.get_hooks("after_delete_page"):
+                pages_to_delete.remove(page)
+                for _page in pages_to_delete:
+                    fn(request, _page)
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:64]: Going to backtrack on this a bit. Sorry @KalobTaulien 

as per https://docs.wagtail.org/en/stable/reference/hooks.html these hooks only apply for the page on which the action is performed via the UI. in this case that is the source page. So this bit is extra 🤦‍♂️😂

File: wagtail/admin/views/pages/delete.py (hunk 3 above)
Comment (zerolab) [reply wagtail/admin/views/pages/delete.py:61]: Hmm... that should be fine as that happens before all this logic

File: wagtail/admin/views/pages/delete.py [hunk 9]
```diff
@@ -50,5 +78,10 @@ def delete(request, page_id):
             "page": page,
             "descendant_count": page.get_descendant_count(),
             "next": next_url,
+            "translation_count": len(pages_to_delete),
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:81]: as in the other case. The confirmation message specifically refers to translations. An alias page is not a translation but a direct copy which gets removed.

as such this should be 

```python
"translation_count": pages_to_delete.filter(alias_of__isnull=True).exclude(page_to_delete).count(),
```

File: wagtail/admin/views/pages/delete.py [hunk 10]
```diff
@@ -50,5 +78,10 @@ def delete(request, page_id):
...
+            "translation_descendant_count": sum(
+                [p.get_descendants().count() for p in pages_to_delete]
+            ),
+            "combined_subpages": page.get_descendant_count() + len(pages_to_delete),
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:85]: ditto for combined subpages. we're only talking about translated (ie. non-alias) pages

**Review by zerolab — CHANGES_REQUESTED** [review]
Thanks for this @KalobTaulien 
Further suggestions for improvement.

This also seems to lack tests 🤔

File: wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html (hunk 7 above)
Comment (KalobTaulien) [reply wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html:17]: Hmm.. Good point, I didn't think about translation context. Yeah let's do what you suggested, that seems like the easiest way forward here.

File: wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html [hunk 11]
```diff
@@ -9,11 +9,24 @@
     <div class="nice-padding">
         <p>
             {% trans 'Are you sure you want to delete this page?' %}
-            {% if descendant_count %}
-                {% blocktrans count counter=descendant_count %}
-                    This will also delete one more subpage.
+
+            {% if descendant_count and translation_count %}
+                {% blocktrans trimmed count counter=translation_count %}
+                    Deleting this page will also delete {{ descendant_count }} more child pages, {{ translation_count }} translation of this page and their combined {{ combined_subpages }} translated child pages.
```
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html:15]: heh, looking at this in practice it _almost_ works. if `descendant_count == 1`, this reads "Deleting this page will also delete 1 more child pages" which is a bit wrong.

```jinja
{% if descendant_count %} {# has descendants #}
    {% blocktrans trimmed count counter=descendant_count %}
    Deleting this page will also delete {{ descendant_count }} child page.
    {% plural %}
    Deleting this page will also delete {{ descendant_count }} more child pages.
    {% endblocktrans %}

    {% if translation_count %} {# has translations #}
        {% if translation_descendant_count %} {# has translations with descendants #}
            {% if translation_count == 1 %}
                {% blocktrans trimmed count counter=translation_descendant_count %}
                It will also delete 1 translation and its combined {{ translation_descendant_count }} translated child page.
                {% plural %}
                It will also delete 1 translation and its combined {{ translation_descendant_count }} translated child pages.
                {% endblocktrans %}
            {% else %}
                {% blocktrans trimmed count counter=translation_descendant_count %}
                It will also delete {{ translation_count }} translations and their combined {{ translation_descendant_count }} translated child page.
                {% plural %}
                It will also delete {{ translation_count }} translations and their combined {{ translation_descendant_count }} translated child pages.
                {% endblocktrans %}
            {% endif %}
        {% else %} {# has translations but no descendants #}
            {% blocktrans trimmed count counter=translation_count %}
            It will also delete {{ translation_count }} translation.
            {% plural %}
            It will also delete {{ translation_count }} translations.
            child pages.
            {% endblocktrans %}
        {% endif %}
    {% endif %}
{% elif translation_count %} {# no descendants #}
    {% if translation_descendant_count %} {# has translations with descendants #}
        {% if translation_count == 1 %}
            {% blocktrans trimmed count counter=translation_descendant_count %}
            Deleting this page will also delete 1 translation and its combined {{ translation_descendant_count }} translated child page.
            {% plural %}
            Deleting this page will also delete 1 translation and its combined {{ translation_descendant_count }} translated child pages.
            {% endblocktrans %}
        {% else %}
            {% blocktrans trimmed count counter=translation_descendant_count %}
            Deleting this page will also delete {{ translation_count }} translations and their combined {{ translation_descendant_count }}
            translated child page.
            {% plural %}
            Deleting this page will also delete {{ translation_count }} translations and their combined {{ translation_descendant_count }}
            translated child pages.
            {% endblocktrans %}
        {% endif %}
    {% else %}
        {% blocktrans trimmed count counter=translation_count %}
        Deleting this page will also delete {{ translation_count }} translation of this page.
        {% plural %}
        This will also delete {{ descendant_count }} more subpages.
        Deleting this page will also delete {{ translation_count }} translations of this page.
        {% endblocktrans %}
    {% endif %}
{% endif %}
```

is probably closer to the truth (need verifying). I split the component parts into two block as the number of variations was too much. cc @phildexter

File: wagtail/admin/views/pages/delete.py [hunk 12]
```diff
@@ -24,12 +25,38 @@ def delete(request, page_id):
...
+        if getattr(settings, "WAGTAIL_I18N_ENABLED", False):
+            for fn in hooks.get_hooks("construct_synced_page_tree_list"):
+                fn_pages = fn([page], "delete")
+                if fn_pages and isinstance(fn_pages, dict):
+                    for additional_pages in fn_pages.values():
+                        pages_to_delete.update(additional_pages)
+
+        pages_to_delete = list(pages_to_delete)
+        translation_count = len(
+            [p.id for p in pages_to_delete if not p.alias_of and p.id != page.id]
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:41]:
```suggestion
            [p.id for p in pages_to_delete if not p.alias_of_id and p.id != page.id]
```

Also best to move this next to `"translation_count":` bit of the context in `TemplateResponse`

File: wagtail/admin/views/pages/delete.py [hunk 13]
```diff
@@ -50,5 +77,10 @@ def delete(request, page_id):
             "page": page,
             "descendant_count": page.get_descendant_count(),
             "next": next_url,
+            "translation_count": translation_count,
+            "translation_descendant_count": sum(
+                [p.get_descendants().count() for p in pages_to_delete]
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:82]: We are only looking at translated pages as the message is "[..] and their combined N translated child pages". so `translation_descendant_count` needs to be `p.get_descendants().filter(alias_of__isnull=True).count()`

File: wagtail/admin/views/pages/delete.py [hunk 14]
```diff
@@ -50,5 +77,10 @@ def delete(request, page_id):
...
+            ),
+            "combined_subpages": page.get_descendant_count() + translation_count,
```
Comment (zerolab) [inline wagtail/admin/views/pages/delete.py:84]: This is extra. We only provide a message about translated pages, which doesn't include aliases

**Review by zerolab — CHANGES_REQUESTED** [review]
Still needs tests.

And a few other comments. Slowly getting there

File: wagtail/contrib/simple_translation/tests/test_wagtail_hooks.py [hunk 15]
```diff
@@ -181,3 +181,67 @@ def test_other_l10n_pages_were_unpublished(self):
...
+            ),
+            follow=True,
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.context["translation_count"], 1)
+        self.assertEqual(response.context["translation_descendant_count"], 0)
+        self.assertIn(
+            "Deleting this page will also delete 1 translation of this page.",
+            response.content.decode("utf-8"),
+        )
```
[the remark this answers is not in the document]
Comment (KalobTaulien) [reply wagtail/contrib/simple_translation/tests/test_wagtail_hooks.py:249]: Done, and done! Can you let me know if there are better checks I should add here?

File: wagtail/contrib/simple_translation/tests/test_wagtail_hooks.py [hunk 16]
```diff
@@ -245,3 +247,97 @@ def test_delete_confirmation_template(self):
...
+
+        # Confirm the fr_blog_post parent id matches the new_parent_page id
+        # This confirms is hasn't moved and hasn't been deleted.
+        # to a different location in the tree.
+        self.fr_blog_post.refresh_from_db()
+        self.assertEqual(
+            self.fr_blog_post.get_parent(update=True).id, self.fr_new_parent.id
+        )
+
+    def test_alias_pages_when_deleting_source_page(self):
```
Comment (zerolab) [inline wagtail/contrib/simple_translation/tests/test_wagtail_hooks.py:323]: as per original notes from https://github.com/wagtail/wagtail/issues/7592#issuecomment-1034790403

> ℹ️  In a synced tree scenario, both translated pages and alias pages will get deleted, unlike the current alias page behaviour on Delete, where the alias page gets unliked from its source (i.e. transform into fully fledged pages) and is preserved. Alias pages created in the same tree will continue to behave as they do currently.

so your test covers the existing behaviour, which is great. It needs, however, to also check the following use case:

<table>
<tr>
<th>before deletion</th><th>after deletion</th>

<tr><td>

```
- source tree root
  - source page
- translated tree root
  - alias of source page
``` 
</td><td>

```
- source tree root
  - source page
- translated tree root
``` 
</td></tr></table>

**Comment by zerolab** [comment]
Thank you @KalobTaulien, merged manually in [ce78949](https://github.com/wagtail/wagtail/pull/7984/commits/ce7894934dc96762a8438fdc3060dfca3c4da98a) in #7984
